src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((line, src)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((line, src))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            // Check for keys that have hit their timeout.
+            while let Ok(Async::Ready(Some(expired_key))) = self.timeouts.poll() {
+                self.expired.push_back(expired_key.into_inner());
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // Handle the incoming line we got from `inner`. If the
+                    // handler gave us something - return it, otherwise continue
+                    // with the flow.
+                    if let Some(val) = self.handle_line(line, src) {
+                        return Ok(Async::Ready(Some(val)));
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    // We got `None`, this means the `inner` stream has ended.
+                    // Start flushing all existing data, stop polling `inner`.
+                    self.draining =
+                        Some(self.buffers.drain().map(|(k, v)| (v.into(), k)).collect());
+                }
+                Ok(Async::NotReady) => {
+                    // We didn't get any lines from `inner`, so we just give
+                    // a line from the expired lines queue.
+                    if let Some(key) = self.expired.pop_front() {
+                        if let Some(buffered) = self.buffers.remove(&key) {
+                            return Ok(Async::Ready(Some((buffered.freeze(), key))));
+                        }
+                    }
+
+                    return Ok(Async::NotReady);
+                }
+                Err(()) => return Err(()),
+            };
+        }
+    }
+}
+
+impl<T> LineAgg<T>
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// Handle line, if we have something to output - return it.
+    fn handle_line(&mut self, line: Bytes, src: Filename) -> Option<(Bytes, Filename)> {
+        // Check if we already have the buffered data for the source.
+        if self.buffers.contains_key(&src) {
+            let condition_matched = self.config.condition_pattern.is_match(line.as_ref());
+            match self.config.mode {
+                // All consecutive lines matching this pattern are included in
+                // the group.
+                Mode::ContinueThrough => {
+                    if condition_matched {
+                        let buffered = self
+                            .buffers
+                            .get_mut(&src)
+                            .expect("already asserted key is present");
+                        buffered.extend_from_slice(b"\n");
+                        buffered.extend_from_slice(&line);
```

Comment:
Given this is repeated on each mode, would it make sense to add a tiny new buffer wrapper that can manage this?

Reply:
Yes, it’s one of the potential improvements. Adding dedicated type wrappers for buffer would make sense, but even before that we could just add helper functions. Another similar improvement would be to add a wrapper around `buffers` HashMap and `timeouts` DelayQueue - to provide a more sound interface.
I decided to skip those because I didn’t want to change the approach on this part of the code beyond the scope of adding features to the multi-line support. It can be viewed as technical debt, and while I’d love to spend more time on it - I don’t think that would be a good time, since we’re trying to push 1.0.

Reply:
I made `LineAgg` generic over the `String` - it's an easy thing to go, and it reduces the footprint of the changes if we'd want to alter the keys at the `file` source.